src/backend/utils/mmgr/aset.c

```diff
@@ -233,6 +233,13 @@ typedef struct AllocChunkData
 	void *prev_chunk;
 	void *next_chunk;
 #endif
+#ifdef EXTRA_DYNAMIC_MEMORY_DEBUG
+#define ALLOC_CHUNK_IS_USED(__chunk__) (__chunk__)->sharedHeader
```

Comment:
Relying on this field in definition whether chunk is freed is not safe as it has multiple meanings in different contexts. From comment:
```
	  * SharedChunkHeader stores all the "shared" details
	  * among multiple chunks, such as memoryAccount to charge,
	  * generation of memory account, memory context that owns this
	  * chunk etc. However, in case of a free chunk, this pointer
	  * actually refers to the next chunk in the free list.
```

The most confident way is to build the set of freed chunks by scanning `freelists` before iteration over chunks to gather statistics. Try to implement this using dynamic hashtable of pointers to freed chunks